sync/src/synchronizer/block_fetcher.rs

```diff
@@ -123,26 +124,43 @@ impl BlockFetcher {
                     break;
                 }
 
-                let to_fetch = self
-                    .snapshot
-                    .get_ancestor(&best_known_header.hash(), index_height)?;
-
-                // NOTE: Filtering `BLOCK_STORED` but not `BLOCK_RECEIVED`, is for avoiding
-                // stopping synchronization even when orphan_pool maintains dirty items by bugs.
-                if self
-                    .snapshot
-                    .contains_block_status(&to_fetch.hash(), BlockStatus::BLOCK_STORED)
-                {
-                    continue;
+                // Try to get [from, to] headers, return the corresponding headers in reversed order.
+                let from = index_height;
+                let to = min(
+                    from + (count - fetch.len()) as u64 - 1,
+                    best_known_header.number(),
+                );
+                let mut candidates = Vec::with_capacity((to - from + 1) as usize);
+                let mut header = self.snapshot.get_ancestor(&best_known_header.hash(), to)?;
+                let mut parent_hash = header.parent_hash();
+
+                candidates.push(header);
+                for _ in from..to {
+                    header = self.snapshot.get_header_view(&parent_hash)?.into_inner();
+                    parent_hash = header.parent_hash();
+                    candidates.push(header);
                 }
 
-                if inflight.insert(self.peer, to_fetch.hash()) {
-                    trace!(
-                        "[Synchronizer] inflight insert {:?}------------{}",
-                        to_fetch.number(),
-                        to_fetch.hash(),
-                    );
-                    fetch.push(to_fetch.hash());
+                index_height = to;
```

Comment:
Please simplify the logic to update `index_height`.

1. On init, set it to `fixed_last_common_header.number()`.
2. Update it in the top of the loop:

    ```
    if index_height > best_known_header.number() {
      break;
    }

    let from = index_height;
    let to = ...;
    index_height = to + 1;
    ```

Reply:
I have rewrite the code. PLTA again.